pkg/controller/statefulset/stateful_pod_control_test.go

```diff
@@ -849,3 +853,402 @@ func collectEvents(source <-chan string) []string {
+				replicaCount, reserveOrdinals := getStatefulSetReplicasCountAndReserveOrdinals(setClone)
+				pods := make([]*v1.Pod, 0)
+				expectIndex := []int{0, 1, 2, 3, 4}
+				currentIndex := make([]int, 0)
+				for i := 0; i < replicaCount; i++ {
+					if reserveOrdinals.Has(i) {
+						continue
+					}
+					currentIndex = append(currentIndex, i)
+					pods = append(pods, newStatefulSetPod(set, i))
+				}
+				if !reflect.DeepEqual(expectIndex, currentIndex) {
+					t.Fatalf("expect(%v), but get(%v)", expectIndex, currentIndex)
+				}
+				return pods
+			},
+			expectPvcOwnerRef: func(pvcName string) metav1.OwnerReference {
+				sIndex1 := strings.Index(pvcName, "-") + 1
+				podName := pvcName[sIndex1:]
+				sIndex2 := strings.LastIndex(pvcName, "-") + 1
+				index, _ := strconv.Atoi(pvcName[sIndex2:])
+				if index < 5 {
+					return metav1.OwnerReference{
+						APIVersion: "apps.kruise.io/v1beta1",
+						Kind:       "StatefulSet",
+						Name:       "foo",
+					}
+				}
+				return metav1.OwnerReference{
+					APIVersion: "v1",
+					Kind:       "Pod",
+					Name:       podName,
+				}
+			},
+		},
+		{
+			name: "reserveOrdinals is nil, scaleDown=true, whenScaled=Delete, whenDeleted=Delete",
+			getStatefulSet: func() *appsv1beta1.StatefulSet {
+				set := newStatefulSet(3)
+				set.Spec.PersistentVolumeClaimRetentionPolicy = &appsv1beta1.StatefulSetPersistentVolumeClaimRetentionPolicy{
+					WhenDeleted: appsv1beta1.DeletePersistentVolumeClaimRetentionPolicyType,
+					WhenScaled:  appsv1beta1.DeletePersistentVolumeClaimRetentionPolicyType,
+				}
+				return set
+			},
+			getPods: func(set *appsv1beta1.StatefulSet) []*v1.Pod {
+				setClone := set.DeepCopy()
+				setClone.Spec.Replicas = utilpointer.Int32(5)
+				replicaCount, reserveOrdinals := getStatefulSetReplicasCountAndReserveOrdinals(setClone)
+				pods := make([]*v1.Pod, 0)
+				expectIndex := []int{0, 1, 2, 3, 4}
+				currentIndex := make([]int, 0)
+				for i := 0; i < replicaCount; i++ {
+					if reserveOrdinals.Has(i) {
+						continue
+					}
+					currentIndex = append(currentIndex, i)
+					pods = append(pods, newStatefulSetPod(set, i))
+				}
+				if !reflect.DeepEqual(expectIndex, currentIndex) {
+					t.Fatalf("expect(%v), but get(%v)", expectIndex, currentIndex)
+				}
+				return pods
+			},
+			expectPvcOwnerRef: func(pvcName string) metav1.OwnerReference {
+				sIndex1 := strings.Index(pvcName, "-") + 1
+				podName := pvcName[sIndex1:]
+				sIndex2 := strings.LastIndex(pvcName, "-") + 1
+				index, _ := strconv.Atoi(pvcName[sIndex2:])
+				if index < 3 {
+					return metav1.OwnerReference{
+						APIVersion: "apps.kruise.io/v1beta1",
+						Kind:       "StatefulSet",
+						Name:       "foo",
+					}
+				}
+				return metav1.OwnerReference{
+					APIVersion: "v1",
+					Kind:       "Pod",
+					Name:       podName,
+				}
+			},
+		},
+		{
+			name: "reserveOrdinals is [2,3], scaleDown=false, whenScaled=Retain, whenDeleted=Delete",
```

Comment:
what about make reserveOrdinals cover the boundary case,e.g. "reserveOrdinals is [2,4]"